com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/StepRequestHandler.java

```diff
@@ -56,63 +67,109 @@ public StepRequestHandler() {
-            ThreadReference thread = DebugUtility.getThread(context.getDebugSession(), threadId);
-            if (thread != null) {
-                setPendingStepKind(thread, command);
-                setOriginalStackDepth(thread);
-                setOriginalStepLocation(thread);
+            context.getDebugSession().getEventHub().events()
+                .filter(debugEvent -> debugEvent.event instanceof StepEvent
+                        || debugEvent.event instanceof BreakpointEvent
+                        || debugEvent.event instanceof ThreadDeathEvent)
+                .subscribe(debugEvent -> {
+                    handleDebugEvent(debugEvent, context.getDebugSession(), context);
+                });
+            return CompletableFuture.completedFuture(response);
+        }
+
+        long threadId = ((StepArguments) arguments).threadId;
+        ThreadReference thread = DebugUtility.getThread(context.getDebugSession(), threadId);
+        if (thread != null) {
+            try {
+                setPendingStepKind(threadId, command);
+                setOriginalStackDepth(threadId, thread.frameCount());
+                setOriginalStepLocation(threadId, getTopFrame(thread).location());
+                StepRequest stepRequest;
                 if (command == Command.STEPIN) {
-                    DebugUtility.stepInto(thread, context.getDebugSession().getEventHub(), context.getDebugFilters().stepFilters);
+                    stepRequest = DebugUtility.stepInto(thread, context.getDebugSession().getEventHub(), context.getDebugFilters().stepFilters);
                 } else if (command == Command.STEPOUT) {
-                    DebugUtility.stepOut(thread, context.getDebugSession().getEventHub(), context.getDebugFilters().stepFilters);
+                    stepRequest = DebugUtility.stepOut(thread, context.getDebugSession().getEventHub(), context.getDebugFilters().stepFilters);
                 } else {
-                    DebugUtility.stepOver(thread, context.getDebugSession().getEventHub(), context.getDebugFilters().stepFilters);
+                    stepRequest = DebugUtility.stepOver(thread, context.getDebugSession().getEventHub(), context.getDebugFilters().stepFilters);
                 }
+                setPendingStepRequest(threadId, stepRequest);
                 ThreadsRequestHandler.checkThreadRunningAndRecycleIds(thread, context);
+            } catch (IncompatibleThreadStateException ex) {
+                final String failureMessage = String.format("Failed to step because the thread %d is not suspended in the target VM.", threadId);
+                logger.log(Level.SEVERE, failureMessage);
+                return AdapterUtils.createAsyncErrorResponse(response, ErrorCode.STEP_FAILURE, failureMessage);
+            } catch (IndexOutOfBoundsException ex) {
+                final String failureMessage = String.format("Failed to step because the thread %d doesn't contain any stack frame", threadId);
```

Comment:
When would this happen?

Reply:
At first, thread.frame(i) claims it could throw IndexOutOfBoundsException. And i found a special case that could get this exception. That are system thread `Attach Listener` and `Signal Dispatcher`, when pausing them, you could find their stack frames are empty.

![image](https://user-images.githubusercontent.com/14052197/33297777-d77d87ca-d41d-11e7-8f46-4ca70a0290dd.png)